File: 105_construct_binary_tree_from_preorder_and_inorder_traversal.cpp
#include <bits/stdc++.h>
#include "package.h"
class Solution{
    public:
        TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder){
            return helper(preorder, 0, preorder.size(), inorder, 0, inorder.size());
        }
        TreeNode* helper(vector<int>& preorder, int i, int j,
                vector<int>& inorder, int ii, int jj){
            if(i>=j && ii>=jj)
                return nullptr;
            TreeNode* root = new TreeNode(preorder[i]);
            auto iter = find(inorder.begin()+ii, inorder.end()+jj, preorder[ii]);
            int len=iter-inorder.begin()-ii;
            root->left=helper(preorder, i+1, i+1+len, inorder, ii, ii+len);
            root->right=helper(preorder, i+1+len, j, inorder, ii+len+1, jj);
            return root;
        }
};

File: 105_construct_binary_tree_from_preorder_and_inorder_traversal.py
from package import TreeNode

class Solution:
    def buildTree(self, preorder, inorder):
        if len(preorder) == 0 and len(inorder) == 0:
            return None
        root = TreeNode(preorder[0])
        idx = root.index(preorder[0])
        root.left = self.buildTree(preorder[1:idx+1], inorder[:idx])
        root.right = self.buildTree(preorder[idx+1:], inorder[idx+1:])
